Add camayoc user documentation

kdelee opened the pull request:
Update README.rst with install information as well as how to run the test suite.
Also provide example pytest.ini and camayoc.yaml config files.

Closes #187

kdelee committed:
Add camayoc user documentation

Update README.rst with install information as well as how to run the test suite.
Also provide example pytest.ini and camayoc.yaml config files.

Add make targets for tests and documents this in README.

Closes #187

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+Installation
+^^^^^^^^^^^^
+
+Camayoc only supports Python 3.5 and python 3.6, so it is recommended that you
```
`Python 3.5 and python 3.6` --> `Python 3.5 and Python 3.6`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+install Camayoc into a virtual environment. There are several tools available
+for managing virtual environments, below is an example of one way.::
+
+    git clone https://github.com/quipucords/camayoc.git
```
I personally prefer to separate the cloning from the virtual environment and break it into steps 1-N.

You can provide a link to: 
- https://docs.python.org/3/tutorial/venv.html
- http://docs.python-guide.org/en/latest/dev/virtualenvs/

Step 1 clone repo:
<example>
Optional Steps (virtual environment):
<create & source environment>

Step 2 install dependencies
make install-dev
make install

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+``$XDG_CONFIG_HOME/camayoc/config.yaml``. On most systems this will be
+``~/.config/camayoc/config.yaml``.
+
+The config file tells camayoc about the hosts you would like to run tests
```
Need consistency around `camayoc` or `Camayoc`. I'd uppercase it based on Jamie's previous opinions with Rho.

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+where the ``quipucords`` server you want to test if you are running
+``quipucords`` tests.
+
+Camayoc contains test suites for both quipucords/rho and quipucords/quipucords
```
links to project repos would be a nice touch.

chambridge replied:
Did you want to update these references too?

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+be done in the ``pytest.ini`` file.
+
+Running the test suite
+^^^^^^^^^^^^
```
The `^^` should go to the end of the section title. Also with titles I would tend to uppercase words.
```
Running the Test Suite
^^^^^^^^^^^^^^^^^^^^^^
```

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+Running the test suite
+^^^^^^^^^^^^
+
+While ``camayoc`` contains test for ``rho`` and ``quipucords``, and these share
```
Not to nitpick but I would again focus on consistency of language and format.
Refer to this project as Camayoc no special formatting.
Refer to Rho and Quipucords or **quipucords/rho**, **quipucords/quipucords**. 

Perhaps a brief intro section before the install would be good so you can then use "shortened" references: **rho**, **quipucords**.

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+a config file format, testing both ``rho`` and ``quipucords`` simultaneously is
+not a supported use case.
+
+Camayoc uses ``py.test`` as its test driver. The pytest project had excellent
```
`The pytest project had excellent` --> `The pytest project has excellent`

chambridge replied:
Still need to fix the `had` --> `has` tense in this sentence.

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+environment. Then invoke the main test suite from the root directory of
+``camayco`` with::
+
+    py.test camayoc/tests/rho/
```
I think there is an opportunity to just add a `make` target for simplicity.

`make test-rho`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+that scan all machines listed in the inventory with all available network
+credentials. To run these, you similarly invoke::
+
+    py.test camayoc/tests/remote/rho/
```
Again opportunity for a `make` target

`make test-rho-remote`

Reduces typing and typos.

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+Running tests for ``quipucords`` and ``qpc``
+^^^^^^
+
+To run all the ``quipucords`` tests, first start the ``quipucords`` server and
```
Might be helpful to provide a link or references to Quipucords install docs?

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+in such a way that it is available in your ``$PATH``. Then run all API, CLI,
+and UI tests with::
+
+    py.test camayoc/tests/qcs/
```
Simplify with `make` target

`make test-qcs`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+only point to one of these subdirectories::
+    
+    # for API tests only
+    py.test camayoc/tests/qcs/api
```
Simplify with `make` target

`make test-qcs-api`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+    # for API tests only
+    py.test camayoc/tests/qcs/api
+    # for CLI tests only
+    py.test camayoc/tests/qcs/cli
```
Simplify with `make` target

`make test-qcs-cli`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+    # for CLI tests only
+    py.test camayoc/tests/qcs/cli
+    # for UI tests only
+    py.test camayoc/tests/qcs/ui
```
Simplify with `make` target

`make test-qcs-ui`

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+you can select tests based on string matching. For example, to run all tests
+with ``create`` in the name and skip any others, run::
+
+    py.test camayoc/tests/qcs/api -k create
```
We did something similar in our make file:
https://github.com/quipucords/quipucords/blob/master/Makefile#L44

Might be nice to do something along these lines with both  file and name "variables".

chambridge commented on README.rst:
```diff
@@ -8,3 +8,121 @@ A GPL-licensed Python library that facilitates functional testing of quipucords.
+
+Or to select a specific test, type out the whole name of the test.
+
+Contributing
```
Often this section is moved into a **CONTRIBUTING.rst**

For example:
https://github.com/quipucords/quipucords/blob/master/CONTRIBUTING.rst

Although I haven't read the above in a while. 😄

chambridge commented on camayoc/tests/qcs/api/v1/utils.py:
```diff
@@ -4,21 +4,14 @@
 import pprint
 import time
 
-from camayoc import config
```
Were the deletions here due to dead code? Just trying to understand how it fits in with this PR.

kdelee replied:
It came up because I was trying to figure out if any code still used the "sources" section under the "qcs" section in the config file. I figured out that I had actuall removed all use of this section from the API and had failed to remove this dead code. I was hoping to get rid of the section all together from the config file, but it turns out it is still used in some CLI tests. Essentially, this section of the config file is on the chopping block and will hopefully go away soon. I wanted to remove this dead code because I didn't want to document it or confuse people.

elyezer replied:
What about squashing all commits but leaving these changes into a separated commits? I know it is related but I think the changes in this file and the changes on the other files can have a logical separation

chambridge reviewed (changes requested):
Made many "nitpick" remarks on this, but wanted to come at it from the point of view of a complete outsider to all the associated projects.

Content is good though. I could also see doing this in two parts if you like my make target suggestions but want to defer that to a follow on PR.

kdelee commented:
@chambridge take a look see some time at the changes I made. Thanks for all the input :)

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+a config file format, testing both rho_ and quipucords_ simultaneously is
+not a supported use case.
+
+Camayoc uses ``py.test`` as its test driver. The pytest project had excellent
```
s/test driver/test runner

elyezer replied:
s/had/has?

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+
+To run all the tests for rho_, first have rho_ installed and available in
+your ``$PATH``. This can be done by either ``rpm`` install to the system or by
+``pip`` installing rho_ from source into your Camayoc virtual
```
Link to rho install docs?

kdelee replied:
All the references use a link to the rho repo (there is a link defined at the top as `rho_`)

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+    make test-rho
+
+There are some additional tests for rho located in ``camayoc/tests/remote/rho``
+that scan all machines listed in the inventory with all available network
```
I think we can make more explicit about the inventory being the inventory available on the Camayoc's config file

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+
+    make test-rho-remote
+
+Running Tests For quipucords_ and qpc_
```
Reading qpc here kind of confuses me with test-qpc which means testing everything under qcs directory. I think we could leave the section title as `Running Tests For quipucords_`

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To run all the quipucords_ tests, first start the quipucords_ server and
+set up your config file to point to the server. You should also install qpc_
```
`qpc_ [executable/binary/cli] (required for CLI tests)`?

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+    directory  because some tests create temporary files that must be accessed
+    by the server.
+
+To only test the API, CLI, or UI, tell pytest this by extending the path to
```
This need to be rephrased, something like `To only test the API, CLI or UI you can take advantage of the following make targets::`

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+    # for UI tests only
+    make test-qpc-ui
+
+The tests can be further selected by specifying any path or file. Additionally
```
This first phrase does not make sense now that make targets are listed above. Maybe remove it?

elyezer commented on README.rst:
```diff
@@ -1,10 +1,161 @@
+
+    make test-qpc-filter pattern=create
+
+Or to select a specific test, type out the whole name of the test.
```
I think worth adding a note here, specifying the hole name does not ensure a single test to be run, if `test_something` exists in more than one file all tests will be run.

elyezer commented on camayoc/tests/qcs/api/__init__.py:
```diff
@@ -5,10 +5,6 @@
-
-Be sure to specify the port if you are running on a non-standard one.
+your camayoc config file. You can find a sample config file in the root of the
+repository named ``example_camayoc.yaml``.
```
This should be updated if the example config file is renamed to `example_config.yaml`.

elyezer commented on example_camayoc.yaml:
```diff
@@ -0,0 +1,130 @@
+---
```
Leaving this comment as a reminder to rename to `example_config.yaml` if that is the name chosen to be used.

chambridge commented:
Overall the updates look good. Made to comments about items that I think you missed, but they are minor.

elyezer commented on example_camayoc.yaml:
```diff
@@ -0,0 +1,130 @@
+        # this section is being outmoded but is still
+        # currenlty used by some CLI tests
+        # if it is not present, some CLI tests will skip but nothing breaks
+        sources:
```
I feel like we still want to be able to control the sources and here list which hosts (by hostname or even by a given name available on the inventory) should be included.

This will remove the possibility of specifying many hosts and credentials to be used. For example when we want to test if quipucords and/or rho will pick the second credential when the first one is invalid.

I mean having that control will benefit for creating more interesting test scenarios in the future.

elyezer reviewed (changes requested):
Thank you for putting all this together. I left some comments that I think will help on some polishing, I hope you find them useful.

kdelee committed:
Delete dead code

Remove utility functions that are no longer being used by anything and
are not up to date with current API usage.

kdelee commented on Makefile:
```diff
@@ -53,6 +64,32 @@ test-coverage:
 	--cov=camayoc.config --cov=camayoc.exceptions --cov=camayoc.utils \
 	--cov=camayoc.api tests
 
+test-rho:
```
done

elyezer commented on Makefile:
```diff
@@ -1,5 +1,6 @@
 TESTIMONY_TOKENS="caseautomation, casecomponent, caseimportance, caselevel, caseposneg, description, expectedresults, id, requirement, setup, subtype1, subtype2, steps, teardown, testtype, upstream, title"
 TESTIMONY_MINIMUM_TOKENS="id, requirement, caseautomation, caselevel, casecomponent, testtype, caseimportance, upstream"
+PYEST_OPTIONS="--verbose"
```
Missing a T s/PYEST_OPTIONS/PYTEST_OPTIONS

elyezer commented on README.rst:
```diff
@@ -1,10 +1,162 @@
+    make test-rho-remote
+
+Running Tests For quipucords_
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Update the heading to match the same number of characters above

elyezer reviewed (approved):
Final comments and this will be good to be merged.

kdelee committed:
Update makefile and config to use 'qpc'

Adds information to README.rst about how to skip the scans at the
beginning of the test session.

Updates makefile and example config file to use 'qpc' now that we've decided to
use that everywhere.
